File: 46. Permutations.java
// recursion

class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        //List<Integer> temp = new ArrayList<Integer>();
        dfs(res, new ArrayList<Integer>(), nums, new boolean[nums.length]);
        return res;
    }
    
    private void dfs(List<List<Integer>> res, List<Integer> temp, int[] nums, boolean[] used) {
        if (temp.size() == nums.length) {
            res.add(new ArrayList<>(temp));
            return;
        }
        
        for (int i = 0; i < nums.length; i++) {
            if (!used[i]) {
                temp.add(nums[i]);
                used[i] = true;
                dfs(res, temp, nums, used);
                temp.remove(temp.size() - 1);
                used[i] = false;
            }
        }
    }
    
    
}


//non recursion

class Solution {
    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> res = new ArrayList<>();
        List<Integer> first = new ArrayList<>();
        first.add(nums[0]);
        res.add(first);
        
        for(int i = 1; i < nums.length; i++) {
            List<List<Integer>> newRes = new ArrayList<>();
            for (List<Integer> temp : res) {
                int size = temp.size() + 1;
                for (int j = 0; j < size; j++) {
                    List<Integer> item = new ArrayList<>(temp);
                    item.add(j,nums[i]);
                    newRes.add(item);
                }
            }
            res = newRes;
        }
        return res;
    }
}

File: amazon/python/46. Permutations.py
class Solution(object):
    def permute(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        def backtrack(first = 0):
            if first == n:
                output.append(nums[:])
            for i in range(first, n):
                nums[i], nums[first] = nums[first], nums[i]
                backtrack(first + 1)
                nums[i], nums[first] = nums[first], nums[i]
                
                
                
        n = len(nums)
        output = []
        backtrack()
        return output
        
#         def backtrack(first = 0):
#             # if all integers are used up
#             if first == n:  
#                 output.append(nums[:])
#             for i in range(first, n):
#                 # place i-th integer first 
#                 # in the current permutation
#                 nums[first], nums[i] = nums[i], nums[first] # this works in python, or we can write a swap function to implement
#                 # use next integers to complete the permutations
#                 backtrack(first + 1)
#                 # backtrack
#                 nums[first], nums[i] = nums[i], nums[first]
        
#         n = len(nums)
#         output = []
#         backtrack()
#         return output

# backtrack：if you use a variable/list, update it, and this is used in continuous loop/function, then backtrack to convert it back;
# non backtrack: 257. Binary Tree Paths.py: each path is independent from others, so will not be used by other recursive calls. 

    
    ## Time: O(N!)
    ## Space: O(N!)
